File: src/Sekoia.Caching.Widgets/sekoia-composition/Sekoia.Caching.Widgets/components/CacheBuilder/ElementCache.ts
import helpers = require("sekoia/helpers");
import CacheEntry = require("./CacheEntry");

class ElementCache<T> implements Sekoia.Caching.IElementCache<T>
{
    private _cache: ICache<T> = { };

    constructor(private _elementRetriever: (id: string) => Rx.Observable<T>, private _maxAgeInMilliseconds = () => 60 * 1000 /* 15 seconds */)
    {
        helpers.assert.parameterCannotBeNull(_elementRetriever, "_elementRetriever");
    }

    public get(id: string): Rx.Observable<T>
    {
        helpers.assert.parameterCannotBeNullOrWhitespace(id, "id");

        const cacheEntry = this.getCacheEntry(id);
        return cacheEntry.observable;
    }

    public refresh(id: string)
    {
        helpers.assert.parameterCannotBeNullOrWhitespace(id, "id");

        const cacheEntry = this.getCacheEntry(id, false);
        if (!cacheEntry)
            return;

        cacheEntry.refresh();
    }

    public refreshAll()
    {
        for (const p in this._cache)
        {
            if (this._cache.hasOwnProperty(p))
                this._cache[p].refresh();
        }
    }

    public remove(id: string)
    {
        helpers.assert.parameterCannotBeNullOrWhitespace(id, "id");

        const cacheEntry = this.getCacheEntry(id, false);
        if (!cacheEntry)
            return;

        cacheEntry.dispose();
        delete this._cache[id];
    }

    public set(id: string, element: T): Rx.Observable<T>
    {
        helpers.assert.parameterCannotBeNullOrWhitespace(id, "id");
        helpers.assert.parameterCannotBeNull(element, "element");

        const cacheEntry = this.getCacheEntry(id);

        cacheEntry.setCachedValue(element);

        return cacheEntry.observable;
    }

    private getCacheEntry(id: string, createIfMissing: boolean = true): CacheEntry<string, T>
    {
        let entryObservable = this._cache[id];
        if (entryObservable == null && createIfMissing)
        {
            entryObservable = this.createCacheEntry(id);
            this._cache[id] = entryObservable;
        }

        return entryObservable;
    }

    private createCacheEntry(id: string): CacheEntry<string, T>
    {
        return new CacheEntry<string, T>(
            id,
            () =>
            {
                return this._elementRetriever(id)
                    .doOnError(() =>
                    {
                        delete this._cache[id];
                    });
            },
            this._maxAgeInMilliseconds);
    }
}

interface ICache<T>
{
    [id: string]: CacheEntry<string, T>;
}

export = ElementCache;

File: src/Sekoia.Caching/Implementation/ElementCache.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using JetBrains.Annotations;

namespace Sekoia.Caching.Implementation
{
    internal class ElementCache<TElement, TId> : IElementCache<TElement, TId>
        where TElement : class
    {
        private readonly Dictionary<TId, CacheEntry<TElement>> _cache = new Dictionary<TId, CacheEntry<TElement>>();
        private readonly Func<TId, Task<TElement>> _elementRetriever;
        private readonly object _mutex = new object();

        public Func<TimeSpan> MaxAge { get; set; } = () => TimeSpan.FromMinutes(1);

        public ElementCache([NotNull] Func<TId, Task<TElement>> elementRetriever)
        {
            _elementRetriever = elementRetriever ?? throw new ArgumentNullException(nameof(elementRetriever));
        }

        public IObservable<TElement> Get(TId id)
        {
            if (id == null)
                throw new ArgumentNullException(nameof(id));

            var cacheEntry = GetCacheEntry(id);
            return cacheEntry.Observable;
        }

        public void Refresh(TId id)
        {
            if (id == null)
                throw new ArgumentNullException(nameof(id));

            var cacheEntry = GetCacheEntry(id, false);
            cacheEntry?.Refresh();
        }

        public void RefreshAll()
        {
            lock (_mutex)
            {
                foreach (var cacheEntry in _cache.Values)
                    cacheEntry.Refresh();
            }
        }

        public void Remove(TId id)
        {
            if (id == null)
                throw new ArgumentNullException(nameof(id));

            lock (_mutex)
            {
                var cacheEntry = GetCacheEntry(id, false);

                cacheEntry?.Dispose();
                _cache.Remove(id);
            }
        }

        public IObservable<TElement> Set([NotNull] TId id, [NotNull] TElement element)
        {
            if (id == null)
                throw new ArgumentNullException(nameof(id));
            if (element == null)
                throw new ArgumentNullException(nameof(element));

            var cacheEntry = GetCacheEntry(id);

            cacheEntry.SetCachedValue(element);

            return cacheEntry.Observable;
        }

        private CacheEntry<TElement> CreateCacheEntry(TId id)
        {
            return new CacheEntry<TElement>(() => _elementRetriever(id),
                                            MaxAge,
                                            () => Remove(id));
        }

        private CacheEntry<TElement> GetCacheEntry(TId id, bool createIfMissing = true)
        {
            lock (_mutex)
            {
                CacheEntry<TElement> cacheEntry;
                if (!_cache.TryGetValue(id, out cacheEntry) && createIfMissing)
                {
                    cacheEntry = CreateCacheEntry(id);
                    _cache[id] = cacheEntry;
                }

                return cacheEntry;
            }
        }
    }
}
